tensorflow/python/kernel_tests/cholesky_op_test.py

```diff
@@ -88,5 +88,55 @@ def testEmpty(self):
     self._verifyCholesky(np.empty([2, 0, 0]))
 
 
+class CholeskyGradTest(tf.test.TestCase):
+  _backprop_block_size = 32
+
+  def getShapes(self, shapeList):
+    return ((elem, int(np.floor(1.2 * elem))) for elem in shapeList)
+
+  def testSmallMatrices(self):
+    np.random.seed(0)
+    shapes = self.getShapes([1, 2, 10])
+    self.runFiniteDifferences(shapes)
+
+  def testOneBlockMatrices(self):
+    np.random.seed(0)
+    shapes = self.getShapes([self._backprop_block_size + 1])
+    self.runFiniteDifferences(shapes, dtypes=(tf.float32, tf.float64),
+                              scalarTest=True)
+
+  def testTwoBlockMatrixFloat(self):
+    np.random.seed(0)
+    shapes = self.getShapes([2 * self._backprop_block_size + 1])
+    self.runFiniteDifferences(shapes, dtypes=(tf.float32,), scalarTest=True)
+
+  def testTwoBlockMatrixDouble(self):
+    np.random.seed(0)
+    shapes = self.getShapes([2 * self._backprop_block_size + 1])
+    self.runFiniteDifferences(shapes, dtypes=(tf.float64,), scalarTest=True)
+
+  def runFiniteDifferences(self, shapes, dtypes=(tf.float32, tf.float64),
+                           scalarTest=False):
+    with self.test_session(use_gpu=False):
+      for shape in shapes:
+        for dtype in dtypes:
+          if not(scalarTest):
+            x = tf.constant(np.random.randn(shape[0], shape[1]), dtype)
+            K = tf.matmul(x, tf.transpose(x)) / shape[0]  # K is posdef
+            y = tf.cholesky(K)
+          else:  # This is designed to be a faster test for larger matrices.
+            x = tf.constant(np.random.randn(), dtype)
+            R = tf.constant(np.random.randn(shape[0], shape[1]), dtype)
+            e = tf.mul(R, x)
+            K = tf.matmul(e, tf.transpose(e)) / shape[0]  # K is posdef
+            y = tf.reduce_mean(tf.cholesky(K))
```

Comment:
@rmlarsen @vrv The speed problem was coming from the fact that for high dimensional outputs y testing the whole cholesky matrix was prohibitively slow. I fixed this by testing a scalar function of the cholesky matrix for the large test cases. This is a realistic use case in practice (e.g computing the log determinant of a psd matrix).  The whole cholesky test on my desktop now takes 0.7 seconds run directly and a bit slower run from bazel. We still have coverage for the whole matrix case on the smaller test matrices. I've re-squashed this hence why it doesn't appear in the diff.